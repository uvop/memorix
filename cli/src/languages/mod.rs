mod python;
mod rust;
mod typescript;

use crate::{
    flat_schema::FlatValidatedSchema,
    parser::{Export, Language},
    validate::ValidatedSchema,
};
use std::collections::HashMap;

pub fn codegen_per_language(
    export: &Export,
    validated_schema: &ValidatedSchema,
    flat_validated_schema: &FlatValidatedSchema,
) -> HashMap<String, String> {
    export
        .files
        .as_ref()
        .unwrap_or(&vec![])
        .iter()
        .map(|file| {
            let path = file.path.to_string();
            let content = match file.language {
                Language::TypeScript => typescript::codegen(validated_schema),
                Language::Python => python::codegen(flat_validated_schema),
                Language::Rust => rust::codegen(flat_validated_schema),
            };
            (path, content)
        })
        .collect()
}
