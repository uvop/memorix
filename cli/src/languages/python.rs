use crate::{
    flat_schema::{
        FlatValidatedNamespace, FlatValidatedReferenceTypeItemKind, FlatValidatedSchema,
        FlatValidatedTypeItem, TypeObjectItem,
    },
    parser::{Engine, Value, ALL_CACHE_OPERATIONS, ALL_PUBSUB_OPERATIONS, ALL_TASK_OPERATIONS},
};

fn indent(level: usize) -> String {
    "    ".repeat(level)
}
fn flat_type_item_to_code(
    flat_type_item: &FlatValidatedTypeItem,
    schema: &FlatValidatedSchema,
) -> String {
    match flat_type_item {
        FlatValidatedTypeItem::Optional(x) => {
            format!("typing.Optional[{}]", flat_type_item_to_code(x, schema)).to_string()
        }
        FlatValidatedTypeItem::Array(x) => {
            format!("typing.List[{}]", flat_type_item_to_code(x, schema)).to_string()
        }
        FlatValidatedTypeItem::Reference(x) => {
            let (namespace, namespace_names) = x.namespace_indexes.iter().fold(
                (&schema.global_namespace, Vec::<&str>::new()),
                |(current_namespace, mut acc), &i| {
                    let (next_name, next_namespace) = &current_namespace.namespaces[i];
                    acc.push(next_name.as_str());
                    (next_namespace, acc)
                },
            );
            format!(
                "{prefix}{seperator}{name}",
                prefix = namespace_names
                    .iter()
                    .map(|x| stringcase::pascal_case(x))
                    .collect::<Vec<_>>()
                    .join("."),
                seperator = match namespace_names.is_empty() {
                    true => "",
                    false => ".",
                },
                name = match x.kind {
                    FlatValidatedReferenceTypeItemKind::TypeItem(i) => {
                        namespace.flat_type_items[i].0.clone()
                    }
                    FlatValidatedReferenceTypeItemKind::TypeObjectItem(i) => {
                        namespace.type_object_items[i].0.clone()
                    }
                    FlatValidatedReferenceTypeItemKind::Enum(i) => {
                        namespace.enum_items[i].0.clone()
                    }
                }
            )
        }
        FlatValidatedTypeItem::Boolean => "bool".to_string(),
        FlatValidatedTypeItem::String => "str".to_string(),
        FlatValidatedTypeItem::U32 => "int".to_string(),
        FlatValidatedTypeItem::U64 => "int".to_string(),
        FlatValidatedTypeItem::I32 => "int".to_string(),
        FlatValidatedTypeItem::I64 => "int".to_string(),
        FlatValidatedTypeItem::F32 => "float".to_string(),
        FlatValidatedTypeItem::F64 => "float".to_string(),
    }
}

fn value_to_code(value: &Value) -> String {
    match value {
        Value::String(x) => format!("\"{x}\""),
        Value::Env(x) => {
            format!("os.environ[\"{x}\"]")
        }
    }
}

fn type_item_object_to_code(
    name: &str,
    type_item_object: &TypeObjectItem,
    schema: &FlatValidatedSchema,
    level: usize,
) -> String {
    let base_indent = indent(level);
    format!(
        r#"
{base_indent}@dataclass
{base_indent}class {name}(object):
{}
"#,
        type_item_object
            .properties
            .iter()
            .map(|(property_name, flat_type_item)| format!(
                "{base_indent}    {property_name}: {}",
                flat_type_item_to_code(flat_type_item, schema)
            )
            .to_string())
            .collect::<Vec<_>>()
            .join("\n")
    )
    .to_string()
}

fn namespace_to_code(
    namespace: &FlatValidatedNamespace,
    name_tree: Vec<String>,
    schema: &FlatValidatedSchema,
) -> String {
    let mut result = String::from("");
    let base_indent = indent(name_tree.len());

    for (name, values) in &namespace.enum_items {
        result.push_str(&format!(
            r#"
{base_indent}class {name}(str, Enum):
{}

"#,
            values
                .iter()
                .map(|x| format!("{base_indent}    {x} = \"{x}\""))
                .collect::<Vec<_>>()
                .join("\n")
        ));
    }
    for (name, type_item_object) in &namespace.type_object_items {
        result.push_str(&format!(
            "{}\n",
            type_item_object_to_code(name.as_str(), type_item_object, schema, name_tree.len())
        ));
    }
    for (name, flat_type_item) in &namespace.flat_type_items {
        result.push_str(&format!(
            "{base_indent}{name} = {}\n\n",
            flat_type_item_to_code(flat_type_item, schema)
        ));
    }
    for (name, sub_namespace) in &namespace.namespaces {
        result.push_str(&format!(
            r#"{base_indent}class {name}(object):
{namespace_content}
"#,
            name = stringcase::pascal_case(name),
            namespace_content = namespace_to_code(
                sub_namespace,
                name_tree
                    .clone()
                    .into_iter()
                    .chain(std::iter::once(name.clone()))
                    .collect(),
                schema,
            ),
        ));
    }
    if !namespace.cache_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixCache(MemorixCacheAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .cache_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixCacheAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},{options}
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_CACHE_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join(""),
                        options = {
                            let content = [
                                item.ttl.as_ref().map(|x| {
                                    format!(
                                        "{base_indent}                ttl={},",
                                        value_to_code(x)
                                    )
                                }),
                                item.extend_on_get.as_ref().map(|x| {
                                    format!(
                                        "{base_indent}                extend_on_get={},",
                                        value_to_code(x)
                                    )
                                }),
                            ]
                            .into_iter()
                            .flatten()
                            .collect::<Vec<_>>()
                            .join("\n");
                            match content.is_empty() {
                                true => "".to_string(),
                                false => format!(
                                    r#"
{base_indent}            options=MemorixCacheAll.Options(
{content}
{base_indent}            )"#
                                ),
                            }
                        }
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    if !namespace.pubsub_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixPubSub(MemorixPubSubAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .pubsub_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixPubSubAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_PUBSUB_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join("")
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    if !namespace.task_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixTask(MemorixTaskAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .task_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixTaskAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},{options}
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_TASK_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join(""),
                        options = {
                            let content = [item.queue_type.as_ref().map(|x| {
                                format!(
                                    "{base_indent}                queue_type={},",
                                    value_to_code(x)
                                )
                            })]
                            .into_iter()
                            .flatten()
                            .collect::<Vec<_>>()
                            .join("\n");
                            match content.is_empty() {
                                true => "".to_string(),
                                false => format!(
                                    r#"
{base_indent}            options=MemorixTaskAll.Options(
{content}
{base_indent}            )"#
                                ),
                            }
                        }
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    let pre_namespace = match name_tree.is_empty() {
        true => "".to_string(),
        false => format!(
            "{}.",
            name_tree
                .iter()
                .map(|x| stringcase::pascal_case(x))
                .collect::<Vec<_>>()
                .join(".")
        ),
    };
    result.push_str(&format!(
        r#"{base_indent}class Memorix(MemorixBase):
{init_def}

{base_indent}        self._namespace_name_tree = [{name_tree}]
{class_content}
"#,
        name_tree = name_tree
            .iter()
            .map(|x| format!("\"{}\"", x))
            .collect::<Vec<_>>()
            .join(", "),
        init_def = match name_tree.is_empty() {
            true => format!(
                r#"{base_indent}    def __init__(self) -> None:
{base_indent}        super().__init__(redis_url={redis_url})
"#,
                redis_url = match &schema.engine {
                    Engine::Redis(x) => value_to_code(x),
                }
            ),
            false => format!(
                r#"{base_indent}    def __init__(self, ref: MemorixBase) -> None:
{base_indent}        super().__init__(ref=ref)
"#
            ),
        },
        class_content = [
            namespace
                .namespaces
                .iter()
                .map(|(name, _)| format!(
                    "{base_indent}        self.{name} = {pre_namespace}{namespace_name}.Memorix(ref=self)",
                    namespace_name = stringcase::pascal_case(name)
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            [
                (!namespace.cache_items.is_empty())
                    .then(|| format!("{base_indent}        self.cache = {pre_namespace}MemorixCache(self)")),
                (!namespace.pubsub_items.is_empty())
                    .then(|| format!("{base_indent}        self.pubsub = {pre_namespace}MemorixPubSub(self)")),
                (!namespace.task_items.is_empty())
                    .then(|| format!("{base_indent}        self.task = {pre_namespace}MemorixTask(self)")),
            ]
            .into_iter()
            .flatten()
            .collect::<Vec<_>>()
            .join("\n")
        ]
        .into_iter()
        .filter(|x| !x.is_empty())
        .collect::<Vec<_>>()
        .join("\n\n"),
    ));
    result
}

pub fn codegen(schema: &FlatValidatedSchema) -> String {
    format!(
        r#"# flake8: noqa
import typing
import os

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from enum import Enum
from memorix_client_redis import (
    MemorixBase,
    MemorixCacheAll,
    MemorixPubSubAll,
    MemorixTaskAll,
)

{}"#,
        namespace_to_code(&schema.global_namespace, vec![], schema)
    )
    .to_string()
}
