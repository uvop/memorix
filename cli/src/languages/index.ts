import { assertUnreachable } from "src/core/utilities";
import { codegenTs } from "./ts/ts";
import { codegenPython } from "./python/python";

export enum Languages {
  typescript = "typescript",
  python = "python",
}

type CodegenByLanguageFn = (schema: string, language: Languages) => string;

export const codegenByLanguage: CodegenByLanguageFn = (scope, language) => {
  switch (language) {
    case Languages.typescript:
      return codegenTs(scope);
    case Languages.python:
      return codegenPython(scope);
    default:
      assertUnreachable(language);
      return "";
  }
};
