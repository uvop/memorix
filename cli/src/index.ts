#!/usr/bin/env node
import yargs from "yargs/yargs";
import { hideBin } from "yargs/helpers";
import { Languages } from "./languages";
import { codegen } from "./codgen";

yargs(hideBin(process.argv))
  .command(
    "codegen <schemaFilePath> <...files>",
    "Codegen memorix schema to code",
    (b) => {
      return b
        .positional("schemaFilePath", {
          describe: "Memorix schema file",
        })
        .positional("files", {
          describe: "Pairs of language to codegen and destination file",
        });
    },
    (argv) => {
      const { schemaFilePath, files } = argv as {
        schemaFilePath: string;
        files: string[];
      };
      const parsedFiles: Parameters<typeof codegen>[0]["files"] = [];
      for (let index = 0; index < files.length; index += 2) {
        const languageStr = files[index];
        const dist = files[index + 1];
        const language = Languages[languageStr];
        if (language === undefined) {
          throw new Error(
            `Unknown language, got "${languageStr}", expected one of "${Object.keys(
              Languages
            ).join(", ")}"`
          );
        }
        if (dist === undefined) {
          throw new Error(`Must pass path for "${language}" codegen`);
        }
        parsedFiles.push({
          dist,
          language,
        });
      }
      if (parsedFiles.length === 0) {
        throw new Error(`Didn't get any languages to codegen`);
      }
      codegen({
        schemaFilePath,
        files: parsedFiles,
      });
    }
  )
  .option("verbose", {
    alias: "v",
    type: "boolean",
    description: "Run with verbose logging",
  })
  .demandCommand(1)
  .parse();
