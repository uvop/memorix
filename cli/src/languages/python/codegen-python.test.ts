import { Languages } from "src/languages";
import { codegen } from "src/codgen";
import fs from "fs";

jest.mock("fs", () => {
  return {
    promises: {
      writeFile: jest.fn(),
      readFile: jest.fn(),
    },
  };
});

const codegenPython: (
  schema: string,
  moreConfig?: string
) => Promise<string> = (schema, moreConfig) =>
  new Promise((res) => {
    const schemaWithOutput = `
  Config {
    output: {
      language: "${Languages.python}"
      file: "generated.ts"
    }
    ${moreConfig}
  }

  ${schema}
  `;

    (fs.promises.readFile as jest.Mock).mockImplementation(async () => ({
      async toString() {
        return schemaWithOutput;
      },
    }));
    (fs.promises.writeFile as jest.Mock).mockImplementation(
      async (path, content) => {
        res(content);
      }
    );

    codegen({ schemaFilePath: "schema.memorix" });
  });

describe("python codegen", () => {
  describe("model", () => {
    it("can generate from model", async () => {
      expect(
        await codegenPython(
          `
            Model User {
                id: int
                name: string
                age: float?
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate from 2 models", async () => {
      expect(
        await codegenPython(
          `
            Model User1 {
                id: int
            }
            Model User2 {
                name: string
                age: int?
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model within model", async () => {
      expect(
        await codegenPython(
          `
            Model User {
                id: int
                papa: {
                    name: string?
                }
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model within model within a model", async () => {
      expect(
        await codegenPython(
          `
            Model User {
                id: int
                papa: {
                    name: string?
                    mama: {
                        sick: boolean
                    }?
                    age: int
                }
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model with array", async () => {
      expect(
        await codegenPython(
          `
            Model User {
                id: int
                names: [string]
                children: [{
                  id: int
                  name: string?
                }?]?
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("throws if model appears twice", async () => {
      expect(
        await codegenPython(
          `
            Model User {
                id: int
            }
            Model User {
                id: int
            }
          `
        )
      ).toThrowError();
    });
  });
  describe("cache", () => {
    it("can generate with inline types", async () => {
      expect(
        await codegenPython(
          `
            Cache {
              user {
                key: int
                payload: string
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate with inline types and convert to CamelCase", async () => {
      expect(
        await codegenPython(
          `
            Cache {
              user_by_number {
                key: int
                payload: {
                  name: string
                  age: int
                }
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can have options", async () => {
      expect(
        await codegenPython(
          `
            Cache {
              user {
                key: int
                payload: string
                options: {
                  expire: {
                    value: 5
                    extendOnGet: true
                  }
                }
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate with inline object type", async () => {
      expect(
        await codegenPython(
          `
            Cache {
              user {
                key: int
                payload: {
                  name: string
                  age: int?
                }?
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate with no key", async () => {
      expect(
        await codegenPython(
          `
            Cache {
              user {
                payload: {
                  name: string
                  age: int?
                }?
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate also with model", async () => {
      expect(
        await codegenPython(
          `
          Cache {
            adminId {
                payload: string?
            }
            user {
                key: string
                payload: User
            }
          }
        
          Model User {
            name: string
            age: int?
          }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("pubsub", () => {
    it("can generate with inline types", async () => {
      expect(
        await codegenPython(
          `
            PubSub {
              message {
                key: int
                payload: string
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("task", () => {
    it("can generate with inline types", async () => {
      expect(
        await codegenPython(
          `
            Task {
              doIt {
                key: int
                payload: string
                returns: boolean
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can have options", async () => {
      expect(
        await codegenPython(
          `
            Task {
              doIt {
                key: int
                payload: string
                returns: boolean
                options: {
                  takeNewest: true
                }
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("enum", () => {
    it("can generate", async () => {
      expect(
        await codegenPython(
          `
            Enum Animals {
              dog
              cat
              person
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("config", () => {
    it("can generate", async () => {
      expect(
        await codegenPython(
          `
            DefaultOptions {
              cache: {
                expire: {
                  value: 5
                  extendOnGet: true
                }
              }
              task: {
                takeNewest: true
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can be null", async () => {
      expect(
        await codegenPython(
          `
            DefaultOptions {
              cache: {
                expire: null
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("namespace", () => {
    it("can generate", async () => {
      expect(
        await codegenPython(
          `
            Namespace user {
              DefaultOptions {
                cache: {
                  expire: {
                    value: 5
                  }
                }
              }
              Cache {
                bio {
                  payload: string
                }
              }
            }
            DefaultOptions {
              cache: {
                expire: {
                  value: 6
                }
              }
            }
            Cache {
              favoriteUser {
                payload: string
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
});
