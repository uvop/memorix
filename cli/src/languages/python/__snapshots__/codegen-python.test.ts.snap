// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`python codegen cache can generate also with model 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
    MemorixClientApi,
    MemorixClientApiDefaults as _MemorixClientApiDefaults,
    MemorixClientCacheApi,
    MemorixClientCacheApiItem,
    MemorixClientCacheApiItemNoKey,
    MemorixClientCacheSetOptions as _MemorixClientCacheSetOptions,
    MemorixClientCacheSetOptionsExpire as _MemorixClientCacheSetOptionsExpire,
)


MemorixClientApiDefaults = _MemorixClientApiDefaults
MemorixClientCacheSetOptions = _MemorixClientCacheSetOptions
MemorixClientCacheSetOptionsExpire = _MemorixClientCacheSetOptionsExpire


@dataclass
class User(object):
    name: str
    age: typing.Optional[int]


class MemorixCacheApi(MemorixClientCacheApi):
    def __init__(self, api: MemorixClientApi) -> None:
        super().__init__(api=api)

        self.adminId = MemorixClientCacheApiItemNoKey[typing.Optional[str]](
            api=self._api,
            id=\\"adminId\\",
            payload_class=typing.Optional[str],
        )
        self.user = MemorixClientCacheApiItem[str, \\"User\\"](
            api=self._api,
            id=\\"user\\",
            payload_class=User,
        )


class MemorixApi(MemorixClientApi):
    def __init__(
        self,
        redis_url: str,
        defaults: typing.Optional[MemorixClientApiDefaults] = None,
    ) -> None:
        super().__init__(redis_url=redis_url, defaults=defaults)

        self.cache = MemorixCacheApi(self)"
`;

exports[`python codegen cache can generate with inline object type 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
    MemorixClientApi,
    MemorixClientApiDefaults as _MemorixClientApiDefaults,
    MemorixClientCacheApi,
    MemorixClientCacheApiItem,
    MemorixClientCacheApiItemNoKey,
    MemorixClientCacheSetOptions as _MemorixClientCacheSetOptions,
    MemorixClientCacheSetOptionsExpire as _MemorixClientCacheSetOptionsExpire,
)


MemorixClientApiDefaults = _MemorixClientApiDefaults
MemorixClientCacheSetOptions = _MemorixClientCacheSetOptions
MemorixClientCacheSetOptionsExpire = _MemorixClientCacheSetOptionsExpire


@dataclass
class CacheUserPayload(object):
    name: str
    age: typing.Optional[int]


class MemorixCacheApi(MemorixClientCacheApi):
    def __init__(self, api: MemorixClientApi) -> None:
        super().__init__(api=api)

        self.user = MemorixClientCacheApiItem[int, typing.Optional[\\"CacheUserPayload\\"]](
            api=self._api,
            id=\\"user\\",
            payload_class=typing.Optional[CacheUserPayload],
        )


class MemorixApi(MemorixClientApi):
    def __init__(
        self,
        redis_url: str,
        defaults: typing.Optional[MemorixClientApiDefaults] = None,
    ) -> None:
        super().__init__(redis_url=redis_url, defaults=defaults)

        self.cache = MemorixCacheApi(self)"
`;

exports[`python codegen cache can generate with inline types 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
    MemorixClientApi,
    MemorixClientApiDefaults as _MemorixClientApiDefaults,
    MemorixClientCacheApi,
    MemorixClientCacheApiItem,
    MemorixClientCacheApiItemNoKey,
    MemorixClientCacheSetOptions as _MemorixClientCacheSetOptions,
    MemorixClientCacheSetOptionsExpire as _MemorixClientCacheSetOptionsExpire,
)


MemorixClientApiDefaults = _MemorixClientApiDefaults
MemorixClientCacheSetOptions = _MemorixClientCacheSetOptions
MemorixClientCacheSetOptionsExpire = _MemorixClientCacheSetOptionsExpire


class MemorixCacheApi(MemorixClientCacheApi):
    def __init__(self, api: MemorixClientApi) -> None:
        super().__init__(api=api)

        self.user = MemorixClientCacheApiItem[int, str](
            api=self._api,
            id=\\"user\\",
            payload_class=str,
        )


class MemorixApi(MemorixClientApi):
    def __init__(
        self,
        redis_url: str,
        defaults: typing.Optional[MemorixClientApiDefaults] = None,
    ) -> None:
        super().__init__(redis_url=redis_url, defaults=defaults)

        self.cache = MemorixCacheApi(self)"
`;

exports[`python codegen cache can generate with no key 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
    MemorixClientApi,
    MemorixClientApiDefaults as _MemorixClientApiDefaults,
    MemorixClientCacheApi,
    MemorixClientCacheApiItem,
    MemorixClientCacheApiItemNoKey,
    MemorixClientCacheSetOptions as _MemorixClientCacheSetOptions,
    MemorixClientCacheSetOptionsExpire as _MemorixClientCacheSetOptionsExpire,
)


MemorixClientApiDefaults = _MemorixClientApiDefaults
MemorixClientCacheSetOptions = _MemorixClientCacheSetOptions
MemorixClientCacheSetOptionsExpire = _MemorixClientCacheSetOptionsExpire


@dataclass
class CacheUserPayload(object):
    name: str
    age: typing.Optional[int]


class MemorixCacheApi(MemorixClientCacheApi):
    def __init__(self, api: MemorixClientApi) -> None:
        super().__init__(api=api)

        self.user = MemorixClientCacheApiItemNoKey[typing.Optional[\\"CacheUserPayload\\"]](
            api=self._api,
            id=\\"user\\",
            payload_class=typing.Optional[CacheUserPayload],
        )


class MemorixApi(MemorixClientApi):
    def __init__(
        self,
        redis_url: str,
        defaults: typing.Optional[MemorixClientApiDefaults] = None,
    ) -> None:
        super().__init__(redis_url=redis_url, defaults=defaults)

        self.cache = MemorixCacheApi(self)"
`;

exports[`python codegen config can be null 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
    MemorixClientApi,
    MemorixClientApiDefaults as _MemorixClientApiDefaults,
)


MemorixClientApiDefaults = _MemorixClientApiDefaults


class MemorixApi(MemorixClientApi):
    def __init__(
        self,
        redis_url: str,
        defaults: typing.Optional[MemorixClientApiDefaults] = None,
    ) -> None:
        schema_defaults=MemorixClientApiDefaults(
            cache_set_options=MemorixClientCacheSetOptions(
                expire=None,
            ),
        )
        merged_defaults = schema_defaults + defaults
        super().__init__(redis_url=redis_url, defaults=merged_defaults)"
`;

exports[`python codegen config can generate 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
    MemorixClientApi,
    MemorixClientApiDefaults as _MemorixClientApiDefaults,
)


MemorixClientApiDefaults = _MemorixClientApiDefaults


class MemorixApi(MemorixClientApi):
    def __init__(
        self,
        redis_url: str,
        defaults: typing.Optional[MemorixClientApiDefaults] = None,
    ) -> None:
        schema_defaults=MemorixClientApiDefaults(
            cache_set_options=MemorixClientCacheSetOptions(
                expire=MemorixClientCacheSetOptionsExpire(
                    value=5,
                ),
            ),
        )
        merged_defaults = schema_defaults + defaults
        super().__init__(redis_url=redis_url, defaults=merged_defaults)"
`;

exports[`python codegen enum can generate 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from enum import Enum
from memorix_client_redis import (
,
)





class Animals(str, Enum):
    dog = \\"dog\\"
    cat = \\"cat\\"
    person = \\"person\\""
`;

exports[`python codegen model can generate from 2 models 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
,
)





@dataclass
class User1(object):
    id: int


@dataclass
class User2(object):
    name: str
    age: typing.Optional[int]"
`;

exports[`python codegen model can generate from model 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
,
)





@dataclass
class User(object):
    id: int
    name: str
    age: typing.Optional[float]"
`;

exports[`python codegen model can generate from model with array 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
,
)





@dataclass
class UserChildren(object):
    id: int
    name: typing.Optional[str]


@dataclass
class User(object):
    id: int
    names: typing.List[str]
    children: typing.Optional[typing.List[typing.Optional[\\"UserChildren\\"]]]"
`;

exports[`python codegen model can generate from model within model 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
,
)





@dataclass
class UserPapa(object):
    name: typing.Optional[str]


@dataclass
class User(object):
    id: int
    papa: \\"UserPapa\\""
`;

exports[`python codegen model can generate from model within model within a model 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
,
)





@dataclass
class UserPapaMama(object):
    sick: bool


@dataclass
class UserPapa(object):
    name: typing.Optional[str]
    mama: typing.Optional[\\"UserPapaMama\\"]
    age: int


@dataclass
class User(object):
    id: int
    papa: \\"UserPapa\\""
`;

exports[`python codegen pubsub can generate with inline types 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
    MemorixClientApi,
    MemorixClientApiDefaults as _MemorixClientApiDefaults,
    MemorixClientPubSubApi,
    MemorixClientPubSubApiItem,
    MemorixClientPubSubApiItemNoKey,
)


MemorixClientApiDefaults = _MemorixClientApiDefaults


class MemorixPubSubApi(MemorixClientPubSubApi):
    def __init__(self, api: MemorixClientApi) -> None:
        super().__init__(api=api)

        self.message = MemorixClientPubSubApiItem[int, str](
            api=self._api,
            id=\\"message\\",
            payload_class=str,
        )


class MemorixApi(MemorixClientApi):
    def __init__(
        self,
        redis_url: str,
        defaults: typing.Optional[MemorixClientApiDefaults] = None,
    ) -> None:
        super().__init__(redis_url=redis_url, defaults=defaults)

        self.pubsub = MemorixPubSubApi(self)"
`;

exports[`python codegen task can generate with inline types 1`] = `
"# flake8: noqa
import typing

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from memorix_client_redis import (
    MemorixClientApi,
    MemorixClientApiDefaults as _MemorixClientApiDefaults,
    MemorixClientTaskApi,
    MemorixClientTaskApiItem,
    MemorixClientTaskApiItemNoKey,
    MemorixClientTaskApiItemNoReturns,
    MemorixClientTaskApiItemNoKeyNoReturns,
    MemorixClientTaskDequequeOptions as _MemorixClientTaskDequequeOptions,
)


MemorixClientApiDefaults = _MemorixClientApiDefaults
MemorixClientTaskDequequeOptions = _MemorixClientTaskDequequeOptions


class MemorixTaskApi(MemorixClientTaskApi):
    def __init__(self, api: MemorixClientApi) -> None:
        super().__init__(api=api)

        self.doIt = MemorixClientTaskApiItem[int, str, bool](
            api=self._api,
            id=\\"doIt\\",
            payload_class=str,
            returns_class=bool,
        )


class MemorixApi(MemorixClientApi):
    def __init__(
        self,
        redis_url: str,
        defaults: typing.Optional[MemorixClientApiDefaults] = None,
    ) -> None:
        super().__init__(redis_url=redis_url, defaults=defaults)

        self.task = MemorixTaskApi(self)"
`;
